src/transport/UnauthenticatedSessionTable.h

```diff
@@ -0,0 +1,204 @@
+            return CHIP_ERROR_NO_MEMORY;
+        }
+
+        mEntries.ResetObject(entry, address);
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * Get a session using given address
+     *
+     * @return the peer found, nullptr if not found
+     */
+    CHECK_RETURN_VALUE
+    UnauthenticatedSession * FindEntry(const PeerAddress & address)
+    {
+        UnauthenticatedSession * result = nullptr;
+        mEntries.ForEachActiveObject([&](UnauthenticatedSession * entry) {
+            if (MatchPeerAddress(entry->GetPeerAddress(), address))
+            {
+                result = entry;
+                return false;
+            }
+            return true;
+        });
+        return result;
+    }
+
+    /**
+     * Get a peer given the peer id. If the peer doesn't exist in the cache, allocate a new entry for it.
+     *
+     * @return the peer found or allocated, nullptr if not found and allocate failed.
+     */
+    CHECK_RETURN_VALUE
+    UnauthenticatedSession * FindOrAllocateEntry(const PeerAddress & address)
+    {
+        UnauthenticatedSession * result = FindEntry(address);
+        if (result != nullptr)
+            return result;
+
+        CHIP_ERROR err = AllocEntry(address, result);
+        if (err == CHIP_NO_ERROR)
+        {
+            return result;
+        }
+        else
+        {
+            return nullptr;
+        }
+    }
+
+    /// Mark a session as active
+    void MarkSessionActive(UnauthenticatedSession & entry) { entry.SetLastActivityTimeMs(mTimeSource.GetCurrentMonotonicTimeMs()); }
+
+    /// Allows access to the underlying time source used for keeping track of connection active time
+    Time::TimeSource<kTimeSource> & GetTimeSource() { return mTimeSource; }
+
+private:
+    UnauthenticatedSession * FindLeastRecentUsedEntry()
+    {
+        UnauthenticatedSession * result = nullptr;
+        uint64_t oldestTimeMs           = std::numeric_limits<uint64_t>::max();
+
+        mEntries.ForEachActiveObject([&](UnauthenticatedSession * entry) {
+            if (entry->GetReferenceCount() == 0 && entry->GetLastActivityTimeMs() < oldestTimeMs)
+            {
+                result       = entry;
+                oldestTimeMs = entry->GetLastActivityTimeMs();
+            }
+            return true;
+        });
+
+        return result;
+    }
+
+    static bool MatchPeerAddress(const PeerAddress & a1, const PeerAddress & a2)
+    {
+        if (a1.GetTransportType() != a2.GetTransportType())
+            return false;
+
+        switch (a1.GetTransportType())
+        {
+        case Transport::Type::kUndefined:
+            return false;
+        case Transport::Type::kUdp:
+        case Transport::Type::kTcp:
+            return a1.GetIPAddress() == a2.GetIPAddress() && a1.GetPort() == a2.GetPort() &&
+                (a1.GetInterface() == INET_NULL_INTERFACEID || a2.GetInterface() == INET_NULL_INTERFACEID ||
+                 a1.GetInterface() == a2.GetInterface());
```

Comment:
What do we intend here by ignoring interface if one interface is INET_NULL_INTERFACEID?  In the case of link-local addresses, these are only reachable if you know the interface ID.  So we might not actually have equality for link-local.

I suspect that our code would work fine even without special-casing INET_NULL_INTERFACEID.  

If not and we want INET_NULL_INTERFACEID to enter into this logic, I think we would also need to account for whether the address is link-local.  If it is, then `a1.GetInterface() == INET_NULL_INTERFACEID || a2.GetInterface() == INET_NULL_INTERFACEID` as a part of an equality test can break.

```suggestion
                 a1.GetInterface() == a2.GetInterface();
```

Reply:
Unfortunately it doesn't work in the current stage.

When initializing a pairing session, we are using a peer address with its interface set to null, so a session is created with its interface set to null. While handling the response, the peer address arguments generated from transport layer contains the real interface, not null. So the response packet is unable to match the original session.

After mdns is completed, we can enforce that the peer address is coming from the result of the mdns discover, such that it contains the real interface, the we can remove the null case. I'll add extra comments to explain it.

Reply:
OK, I do see now that INET_NULL_INTERFACEID must be specially accommodated and accounted for.  But there is one case where the above doesn't work.  Link-local addresses are ambiguous without the interface ID.  If both a1 and a2 are link-local and both have INET_NULL_INTERFACEID, you cannot say they are equal even if the IP addresses are the same.  The following is a mouthful (and you could restructure this to make it more readable).  But this would account for this:

```suggestion
                !(a1.GetIPAddress().IsIPv6LinkLocal() && a1.GetInterface() == INET_NULL_INTERFACEID && a2.GetIPAddress().IsIPv6LinkLocal() && a2.GetInterface() == INET_NULL_INTERFACEID) &&
                (a1.GetInterface() == INET_NULL_INTERFACEID || a2.GetInterface() == INET_NULL_INTERFACEID ||
                 a1.GetInterface() == a2.GetInterface())
```

 > When initializing a pairing session, we are using a peer address with its interface set to null

And I'm assuming it must be the case then that this peer address is not a link-local address.  It's impossible that this would work if it was, right?

Reply:
Actually, even that misses a case.  Nodes associated with non-link-local addresses aren't further distinguishable by their interface ID.  Imagine you had two identical non-link-local IP addresses associated to two separate non-null interface IDs.  These don't represent different peers.

This simplifies further and fixes all of these problems.  And this specifically works in the case where interface ID is being scraped from a response, which can only make sense for non-link-local addresses.

```suggestion
                (!a1.GetIPAddress().IsIPv6LinkLocal() || !a2.GetIPAddress().IsIPv6LinkLocal() ||
                 a1.GetInterface() == a2.GetInterface());
```

Reply:
Understand, interface is relevant iff the address is link-local.

Reply:
Yes, exactly.  And your change makes this even clearer.

Thanks!